### ISSUE TITLE
Question about blockwise operation of the algorithm

### ISSUE THREAD

**Issue by TLINDEN** [merrell42/model-synthesis#6]
Hi Paul,

I am about to implement this algorithm in Go. I have the "WFC" part working and now want to implement the blockwise operation part, but I'm not sure if I fully understood it. So, if you allow, could you please tell me if I see it correctly?

As far as I understand it, it has to work like this:

1. Start with an output map (say with 32x32 slots) with each slot set to the superposition containing all modules available
2. Select a block (of say 8x8 slots)
4. Select a random slot (or if there already has been some collapsing happened, select the one with the lowest entropy)
5. Collapse it to one module
6. Propagate to the boundaries of the block
7. Check all slots neighboring the block and reduce the modules in those slots directly neighboring these slots
8. Shift the block by one slot to some arbitrary direction and restart from (4)

Does that make sense? I'm not sure, especially with part (7). When exactly do I have to check the neighboring slot? Before I collapsed one slot inside the block or afterwards?

Also in the paper (and video) you say that the slots inside the block have to be set to the superposition during each run (if I understood it right) - but then how would the algorithm ever come to an end?

I would really appreciate it if you could help me a little!

Thanks in advance,
Tom

**Comment by merrell42** [comment]
Your description is not correct. The first step which I don't see is to find a trivial solution. Find any solution that will work. Often this is just empty space and a ground plane. In my code, I look for tiles that repeat themselves.

The main idea is were going to start with a trivial solution and incrementally modify it. We select different blocks to modify and run the normal model synthesis algorithm on each block. You ask how the algorithm comes to an end. All we are doing is running the normal algorithm on each block we select.

In your step 1, you're consider all possible labels (modules) across the whole output map. We're only going to do that within each block.

We should consider the neighbors of blocks (your step 6) as soon as we select a block (right after step 2). We do that before we start running the original algorithm where we select a label (your step 4) and then propagate it.

In step 3, I prefer going in scanline order, but you can use lowest entropy.

In step 7, I normally move the block by half the width of block. Basically, I pick the blocks in scanline order.

I explain more in my video at 8:47: https://www.youtube.com/watch?v=A2ODauA1a0M&t=536s
And more in my dissertation. See Figure 3.9, Algorithm 3.4, Sec. 3.3.6 (https://paulmerrell.org/wp-content/uploads/2021/06/thesis.pdf).

Let me know if that helps. I'm happy to answer more questions.

**Comment by TLINDEN** [comment]
Thank you very much for your response.

Ok, just to recap:

1. fill in the whole map with a simple tile (which has the same constraint on all 4 sides, I'd say)
2. select a block
3. check all neighbor slots of the block and propagate their constraints towards inside the block
4. then start the actual algorithm, select one random slot (or as in WFC the one with the lowest entropy), collapse
5. propagate
6. move the block by half its size
7. start at 3

As I understand it, I'll be ready when I have covered the whole map with the blocks.

Ok, but the step between 3+4 is still unclear to me: when I create a new block and first look at the neighbors on the blocks edges, they'll be already [partly] collapsed, so I need to check their current modules constraints and reduce the modules of the neighboring slots inside the block accordingly.

Now what? Since I have reduced the edge slots, I'd need to propagate this change throughout the whole block, don't I? If yes, then after collapsing a random slot, I'd need to propagate AGAIN. This can't be right.

I'm actually more confused the more I think about it :)

**Comment by TLINDEN** [comment]
Looked at the video, so you remove the labels of each block, then look at the outside neighbors, propagate the possible models throughout the block and THEN collapse one slot in the block and propagate this change again.

Does that make sense?

**Comment by merrell42** [comment]
Yes, that sounds right to me. We know what all the neighbors should be so we propagate them. And then we start collapsing within the block, we propagate that information. I think maybe you got it now, but let me know if you have more questions.
